packages/lib/transformDateTimeRange.ts

```diff
@@ -0,0 +1,114 @@
+import type { EventType } from "@prisma/client";
+
+import type { Dayjs } from "@calcom/dayjs";
+import dayjs from "@calcom/dayjs";
+import { PeriodType } from "@calcom/prisma/enums";
+
+import { getUTCOffsetByTimezone } from "./dayjs";
+import { calculatePeriodLimits } from "./isOutOfBounds";
+
+export type DateTimeRangeTransformOptions = {
+  eventType: Pick<
+    EventType,
+    | "periodType"
+    | "periodDays"
+    | "periodCountCalendarDays"
+    | "periodStartDate"
+    | "periodEndDate"
+    | "minimumBookingNotice"
+  > & {
+    schedule: {
+      timeZone: string;
+    };
+  };
+  timeZone: string;
+};
+
+export type DateTimeRangeSuccess = {
+  success: true;
+  start: Dayjs;
+  end: Dayjs;
+};
+
+export type DateTimeRangeFailure = {
+  success: false;
+  field: string;
+  message: string;
+};
+
+const getStartTime = (startTimeInput: string, timeZone?: string, minimumBookingNotice?: number) => {
+  const startTimeMin = dayjs.utc().add(minimumBookingNotice || 1, "minutes");
+  const startTime = timeZone === "Etc/GMT" ? dayjs.utc(startTimeInput) : dayjs(startTimeInput).tz(timeZone);
+
+  return startTimeMin.isAfter(startTime) ? startTimeMin.tz(timeZone) : startTime;
+};
+
+const transformDateTimeRange = (
+  startTime: string,
+  endTime: string,
+  { eventType, timeZone }: DateTimeRangeTransformOptions
+): DateTimeRangeSuccess | DateTimeRangeFailure => {
+  const isRollingWindowPeriodType = eventType.periodType === PeriodType.ROLLING_WINDOW;
+  const isStartTimeInPast = dayjs(startTime).isBefore(dayjs().subtract(1, "day").startOf("day"));
+  // If startTime is already sent in the past, we don't need to adjust it.
+  // We assume that the client is already sending startTime as per their requirement.
+  // Note: We could optimize it further to go back 1 month in past only for the 2nd month because that is what we are putting a hard limit at.
+  const startTimeAdjustedForRollingWindowComputation =
+    isStartTimeInPast || !isRollingWindowPeriodType
+      ? startTime
+      : dayjs(startTime).subtract(1, "month").toISOString();
+
+  let startTimeResult = getStartTime(
+    startTimeAdjustedForRollingWindowComputation,
+    timeZone,
+    eventType.minimumBookingNotice
+  );
+
+  let endTimeResult = timeZone === "Etc/GMT" ? dayjs.utc(endTime) : dayjs(endTime).utc().tz(timeZone);
+
+  const eventUtcOffset = getUTCOffsetByTimezone(eventType.schedule.timeZone) ?? 0;
+  const bookerUtcOffset = timeZone ? getUTCOffsetByTimezone(timeZone) ?? 0 : 0;
+  const periodLimits = calculatePeriodLimits({
```

Comment:
I see this is taken from line 626 in `packages\trpc\server\routers\viewer\slots\util.ts`. Do we need to go through the logic of looping through each slot to see if they fall within the period dates if we're already setting the period end date earlier in the function?

Reply:
No; ideally not so but still WIP, writing more tests.